W-21953173-Address-Jack-feedback-CP (#626)
mulesoft/docs-code-builder: modules/ROOT/pages/int-debug-mule-apps.adoc

+
When execution reaches the breakpoint, processing pauses so you can inspect variables, payloads, and the execution state in the Debug panel.

NOTE: Batch execution is concurrent. Multiple record blocks can be processed in parallel, so breakpoints can trigger on different threads. The Debug panel displays each thread context so you can inspect each execution path.

[[conditional-breakpoint]]
== Set a Conditional Breakpoint

> A conditional breakpoint uses a DataWeave expression so execution pauses at that line only when the expression evaluates to `true`.

Reviewer: ```suggestion
A conditional breakpoint uses a DataWeave expression, so execution pauses at that line only when the expression evaluates to `true`.
```

Author: OK! Thanks!

To set or change a breakpoint condition:

. Add a breakpoint in the configuration XML. See <<add-breakpoint>>.
. Right-click the breakpoint indicator in the gutter for that line.
. Select *Edit Breakpoint*.
. In the breakpoint editor, enter a DataWeave expression that uses the current Mule event.

The gutter context menu for a breakpoint also includes *Remove Breakpoint*, *Disable Breakpoint*, and *Run to Line*.

> To review a condition later, hover the breakpoint in the gutter. A tooltip shows the configured expression in the form `Condition:` followed by your DataWeave expression.

Reviewer: ```suggestion
To review a condition later, hover over the breakpoint in the gutter. A tooltip shows the configured expression in the form `Condition:` followed by your DataWeave expression.
```

Author: OK! Thanks!

[[evaluate-dw-while-debugging]]
== Evaluate DataWeave While Debugging

While execution is paused, you can run DataWeave against the current Mule event from the *Debug Console* or from pinned entries in the *Watch* panel. For the variables available in those expressions, see xref:dataweave::dataweave-variables-context.adoc[Predefined Mule event variables].

[[evaluate-dw-debug-console]]
=== Use the Debug Console

. Start a debug session and pause execution at a breakpoint or after a step. See <<run-debug-mode>> and <<debug-toolbar>>.
. Open the *Debug Console* tab, or choose *View* > *Debug Console*.
. Ensure the input is set to *Expression*. Type a DataWeave expression and press Enter to evaluate it against the current message, attributes, variables, and other values visible to the debugger at the pause.
+
The console shows each expression you run and its evaluated result for the current pause.

The *Debug Console* is suited to check if needed as you step through a flow.

[[evaluate-dw-watch]]
=== Use the Watch Panel

. Open the *Watch* panel in the *Run and Debug* sidebar.
. Add an entry with *+* and enter a DataWeave expression.
+

> Whenever execution pauses, the debugger re-evaluates each watch expression and shows the result next to the expression, similar to the *Debug Console*.

Reviewer: ```suggestion
Whenever execution pauses, the debugger reevaluates each watch expression and shows the result next to the expression, similar to the *Debug Console*.
```

Author: OK! Thanks!
